state-commitments/optimized/history_commitment.go

```diff
@@ -20,21 +20,48 @@ type Commitment struct {
 	LastLeaf       common.Hash
 }
 
+// NewCommitment produces a history commitment from a list of leaves that are virtually padded using
+// the last leaf in the list to some virtual length, without making those extra allocations needed to do so.
+// Virtual must be >= len(leaves).
 func NewCommitment(leaves []common.Hash, virtual uint64) (*Commitment, error) {
 	if len(leaves) == 0 {
 		return nil, errors.New("must commit to at least one leaf")
 	}
+	if virtual == 0 {
+		return nil, errors.New("virtual size cannot be zero")
+	}
+	if virtual < uint64(len(leaves)) {
+		return nil, errors.New("virtual size must be greater than or equal to the number of leaves")
+	}
 	comm := NewCommitter()
+	firstLeaf := leaves[0]
+	lastLeaf := leaves[len(leaves)-1]
+	var firstLeafProof, lastLeafProof []common.Hash
+	ok, err := isPowTwo(virtual)
+	if err != nil {
+		return nil, err
+	}
+	if ok {
```

Comment:
`ok` is a bit odd for this variable name, as we don't error if it's false. I think a comment explaining why we only need proofs for the first and last leaves if virtual is a power of two would also be helpful.

Reply:
Cleaned up in #686

Now, when the callers want an inclusion proof with their history commitment, then call `NewCommitment` and when they don't, they just call `ComputeRoot`